src/wallet/coin_selection.rs

```diff
@@ -567,27 +571,28 @@ impl BranchAndBoundCoinSelection {
         for _ in 0..BNB_TOTAL_TRIES {
             // Conditions for starting a backtrack
             let mut backtrack = false;
-            // Cannot possibly reach target with the amount remaining in the curr_available_value,
+            // Cannot possibly reach target with the amount remaining in the available_value,
             // or the selected value is out of range.
             // Go back and try other branch
-            if curr_value + curr_available_value < target_amount
-                || curr_value > target_amount + cost_of_change as i64
+            if selected_value + available_value < target_amount
+                || selected_value > target_amount + cost_of_change as i64
             {
                 backtrack = true;
-            } else if curr_value >= target_amount {
+            } else if selected_value >= target_amount {
                 // Selected value is within range, there's no point in going forward. Start
                 // backtracking
                 backtrack = true;
 
                 // If we found a solution better than the previous one, or if there wasn't previous
                 // solution, update the best solution
-                if best_selection_value.is_none() || curr_value < best_selection_value.unwrap() {
+                if best_selection_value.is_none() || selected_value < best_selection_value.unwrap()
+                {
                     best_selection = current_selection.clone();
-                    best_selection_value = Some(curr_value);
+                    best_selection_value = Some(selected_value);
                 }
```

Comment:
Yes, of course! I realized we passed the `required_utxos` as parameter to all
the algorithms, but in the end, they worked as a fixed value in all coin
selections, and the `required_utxos` should be included no-matter-what in the
CoinSelectionResult, as its original logic.
In BranchAndBound, something similar happened with `curr_value` because
it was the `required_utxos`'s summed amount.
`curr_value`, as implemented at that moment, wasn't a fixed value, but it had a
lower bound: the amount compounded by `required_utxos`. So, my idea was to
remove the `required_utxos` parameter, and related parameters like `curr_value`,
from the `coin_select` signature, and create a new `target_amount` subtracting the
effective value of `required_utxos`. The coin selection would depend from
that moment on the `optional_utxos` and the new target.
`curr_value` followed the same logic with the new target: the lower bound became
zero because we were already considering the `required_utxos`'s amount in the new
`target_amount`, and then there would be no need to take it as a parameter.
The change of the name was because it eased my thinking while making the
changes, I mean, in the end, `curr_value` held the value of all the
`selected_utxos`, so it was the "selected_value".

Reply:
I think we're talking past each other. What I was asking about seems unrelated to the required_utxos to me. The Branch and Bound Coin Selection Strategy aims to find a changeless input set that minimizes a transaction's _waste_ (which is a proxy for the transaction's cost in the context of the wallet's lifetime). To that end, we keep the input set that produces the lowest `curr_waste`, rather than the lowest `selected_value`.

![image](https://user-images.githubusercontent.com/4060799/186671934-4b6b31bf-17de-4cca-8a11-546a3672a11a.png)

The motivation here is that e.g. at high feerates you should prefer using two inputs that drop 100 sats more to fees, instead of paying the fees for five inputs that overshoot the target slightly less.

So, when I saw

```
                if best_selection_value.is_none() || selected_value < best_selection_value.unwrap()
                {
                    best_selection = current_selection.clone();
                    best_selection_value = Some(selected_value);
                }
 ```

 I was wondering whether you were optimizing for the smallest overshoot on purpose, or whether you also wanted to optimize for smallest waste instead, as we do in Bitcoin Core.

Reply:
@Xekyo I think you're right in saying that we're not trying to minimize the waste: my understanding of the bnb coin selection wasn't optimal when I wrote the code, and it hasn't been updated ever since 😰 Thanks for pointing this out!
I opened an issue for this (https://github.com/bitcoindevkit/bdk/issues/733), and we'll try to fix asap :)

On the other end, I don't think it should be fixed in this PR, as, If I understand correctly, here @csralvall  is just refactoring the code structure (changing variable names, such as going from `curr_value` to `selected_value`, trying to abstract some operations with the new `get_selection` function, etc) - being a big enough PR already, I think we can work on fixing the algorithm in a subsequent one :)

Reply:
+1 - I was confused because the logic is the same as in the master branch. This PR is mainly refactoring.

Reply:
Ah, I see. Thanks for explaining